### PR TITLE
feat(warehouse): added support for bigquery custom partition for workspaceIDs

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- With support for Multi-tenancy, we won't be able to support such a config `customPartitionsEnabled` which is at a workspace level. 
- Added support for `customPartitionsEnabledWorkspaceIDs` which will take a list of workspace IDs that needs to have Custom Partition enabled.
- Added support for string slice `checkAndHotReloadConfig`.

### Partitioned tracks table looks like this

Screenshot 2022-11-10 at 3 10 39 PM

### Error experienced without changes

```
2022-11-10T09:29:37.347Z	INFO	warehouse.bigquery	bigquery/bigquery.go:311	BQ: customPartitionsEnabled: bool, customPartitionsEnabledWorkspaceIDs BpLnfgDsc2WD8F2qNfHK5a84jjJ
2022-11-10T09:29:39.702Z	ERROR	warehouse	warehouse/table_upload.go:169	[WH]: Failed uploading table-tracks for upload-1: {Location: "gs://**/**/tracks/24p1HhPk09FW25Kuzxv7GshCLKR/9a7bab75-6b00-400c-941e-441b1a48ba58-tracks/1668072354.24p1HhPk09FW25Kuzxv7GshCLKR.ba545364-17af-4eb4-818c-4a1dc9763ea4.tracks.0a9c6a0a-6523-4b7b-809a-7492bec1b038.json.gz"; Message: "Error while reading data, error message: JSON table encountered too many errors, giving up. Rows: 1; errors: 1. Please look into the errors[] collection for more details. File: gs://**/**/tracks/24p1HhPk09FW25Kuzxv7GshCLKR/9a7bab75-6b00-400c-941e-441b1a48ba58-tracks/1668072354.24p1HhPk09FW25Kuzxv7GshCLKR.ba545364-17af-4eb4-818c-4a1dc9763ea4.tracks.0a9c6a0a-6523-4b7b-809a-7492bec1b038.json.gz"; Reason: "invalid"}
```

## Notion Ticket

https://www.notion.so/rudderstacks/BQ-custom-partition-fix-for-Multi-tenant-9c3262e69c8c4d4f8ecf6f91207e3352

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by Jayachand — APPROVED** [review]

File: warehouse/bigquery/bigquery.go [hunk 1]
```diff
@@ -300,12 +301,12 @@ func (bq *HandleT) loadTable(tableName string, _, getLoadFileLocFromTableUploads
...
 	loadTableByAppend := func() (err error) {
 		stagingLoadTable.partitionDate = time.Now().Format("2006-01-02")
-		outputTable := tableName
+		outputTable := partitionedTable(tableName, stagingLoadTable.partitionDate)
 		// Tables created by RudderStack are ingestion-time partitioned table with pseudo column named _PARTITIONTIME. BigQuery automatically assigns rows to partitions based
 		// on the time when BigQuery ingests the data. To support custom field partitions, omitting loading into partitioned table like tableName$20191221
 		// TODO: Support custom field partition on users & identifies tables
-		if !customPartitionsEnabled {
-			outputTable = partitionedTable(tableName, stagingLoadTable.partitionDate)
+		if customPartitionsEnabled || misc.Contains(customPartitionsEnabledWorkspaceIDs, bq.warehouse.WorkspaceID) {
```
Comment (lvrach) [inline warehouse/bigquery/bigquery.go:308]: You can use [slice.Contains](https://pkg.go.dev/golang.org/x/exp/slices#Contains
)
```suggestion
		if customPartitionsEnabled || slices.Contains(customPartitionsEnabledWorkspaceIDs, bq.warehouse.WorkspaceID) {
```

**Review by Jayachand — APPROVED** [review]

**Review by lvrach — APPROVED** [review]

File: config/load.go [hunk 2]
```diff
@@ -161,6 +162,23 @@ func (c *Config) checkAndHotReloadConfig(configMap map[string][]*configValue) {
...
+					if c.IsSet(key) {
+						isSet = true
+						_value = c.GetStringSlice(key, configVal.defaultValue.([]string))
+						break
+					}
+				}
+				if !isSet {
+					_value = configVal.defaultValue.([]string)
+				}
+				if !reflect.DeepEqual(_value, *value) {
```
Comment (lvrach) [inline config/load.go:178]: Should we do a normal []string comparison instead of reflect?

Comment (achettyiitr) [reply config/load.go:178]: Will this work?
`slices.Compare(_value, *value)`

Comment (achettyiitr) [reply config/load.go]: Using `slices.Compare(_value, *value) != 0` now.

**Pull request merged**